### PR TITLE
Disable product type button when new changes are not saved

### LINKED ISSUE #28738: Change product type and lost my product data
## Describe the bug

As merchant, when I do changes in my product and then change the product. I lost all my data because I didn't save the changes first.

## Expected solution

When the product has unsaved modification, the user can't change the type of the product.
When the product has unsaved modification
Then change type button is disabled.

When the product modification have been saved / There are no unsaved modification
Then change type button is enabled.

## Test

Test for each product type if I can't change the product if there is unsaved modification.

### PR DESCRIPTION
**Pull request by margud**
| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | 8.0.x
| Description?      | In admin product page the product type button is disabled if any inputs where changed.
| Type?             |  improvement 
| Category?         | BO 
| BC breaks?        | no
| Fixed ticket?     | Fixes #28738
| How to test?      | Change any input in product page and the product type button should be disabled.

### CODE REVIEW

File: admin-dev/themes/new-theme/scss/pages/product/product_page.scss
Comment (NeOMakinG) [inline admin-dev/themes/new-theme/scss/pages/product/product_page.scss]:
```suggestion
        color: $gray-500;
        cursor: not-allowed;

        &:hover {
          color: $gray-500;
        }
```

**Review by NeOMakinG — CHANGES_REQUESTED** [review]

**Review by NeOMakinG — DISMISSED** [review]

**Review by matthieu-rolland — DISMISSED** [review]

File: admin-dev/themes/new-theme/js/pages/product/edit/product-type-switcher.ts
Comment (jolelievre) [inline admin-dev/themes/new-theme/js/pages/product/edit/product-type-switcher.ts]: With this it means as soon as something is changed on the page you can't use the button ever again. Even if you revert the modifications. We have a component that checks the product state more accurately here: https://github.com/PrestaShop/PrestaShop/blob/057111abc17f0e0088409f1f829011a2ae43fd0f/admin-dev/themes/new-theme/js/pages/product/edit/product-partial-updater.ts#L226

It's where we toggle the footer buttons' states, so that's where you should toggle the switch component as well. You'll need to pass the `ProductTypeSwitcher` as an argument to the updater, and add a `toggle` method or two enable/disable methods to drive it.

**Review by jolelievre — COMMENTED** [review]
I don't want to block this one since it's already been approved, but I suggested an improvement that would improve the UX.

Also, did you search if it was possible to add some tooltip on hover when it's disabled to explain why? I think maybe using some bootstrap features like `data-bs-toggle="tooltip"` it should be doable.

File: admin-dev/themes/new-theme/js/pages/product/edit/product-type-switcher.ts
Comment (margud) [reply admin-dev/themes/new-theme/js/pages/product/edit/product-type-switcher.ts]: while trying to use this logic noticed that in getUpdatedFormData there is always some info that is seen as changed because its null and is added empty string.
image

in the image you can see what values are seen as changed. Investigating if it works fine if we dont change null to empty string. Not sure if this should be new issue or can be added on this one.

**Review by SiraFily28 — COMMENTED** [review]
Hello 

I noticed that que CI are failed  
Capture d’écran 2022-10-14 à 14 35 34

Ping @PrestaShop/prestashop-core-developers could you please check it ?
Thank you

File: admin-dev/themes/new-theme/js/pages/product/edit/product-partial-updater.ts
Comment (jolelievre) [inline admin-dev/themes/new-theme/js/pages/product/edit/product-partial-updater.ts]: The partial updater has a role of observing the existing components and acting depending on their behavior/interactions. So it can depend on the `$productTypeSwitcher` but it shouldn't be the one to create it. This responsibility should remain in the page form index file as it used to be

File: admin-dev/themes/new-theme/js/pages/product/edit/product-partial-updater.ts
Comment (jolelievre) [inline admin-dev/themes/new-theme/js/pages/product/edit/product-partial-updater.ts]: I'm not sure you should change this, there's an explicit comment that explains why this as needed and it's not in the scope of your PR to change this behavior

This extra filling is required when we need to empty/clear some fields

File: admin-dev/themes/new-theme/js/pages/product/edit/product-type-switcher.ts
Comment (jolelievre) [inline admin-dev/themes/new-theme/js/pages/product/edit/product-type-switcher.ts]: I don't think this should be part of the public API, especially here you disabled/reenable the interactions of two components inside the updater but the updater shouldn't need to know how to handle those interactions

The component itself is responsible for it and could have a public API enable/disable (which is used by the updater), that handles disabling or blocking the click events.

Seems like the preview and the type switcher need to be aware of each other, either via a direct dependency (one is injected into another) or via some events, but the injection is fine I think (since the switcher is also used in another context like creation modal it should remain independent, but it can be injected as a dependency of the preview component however).

**Review by jolelievre — CHANGES_REQUESTED** [review]
Thanks @margud 

A few refactor needed on this one, also you should rebase this one and target develop branch instead (8.1 version), because we won't have time to fix it for the 8.0 release.

**Review by jolelievre — COMMENTED** [review]
Hi @margud 

this PR shouldn't be forgotten. And it also should be rebased to target develop now.

**Comment by kpodemski** [comment]
Hi @margud  @jolelievre 

A little reminder about this PR.

**Comment by jolelievre** [comment]
Hi @margud 

now the PR should target 8.1.x ^^

**Comment by jolelievre** [comment]
@margud the tests are failing and you have conflicts to fix ^^ You probably need to rebase it

**Comment by jolelievre** [comment]
Running the auto campaign https://github.com/jolelievre/ga.tests.ui.pr/actions/runs/4086951694

**Review by jolelievre — APPROVED** [review]

**Review by zuk3975 — APPROVED** [review]

**Review by MhiriFaten — APPROVED** [review]
Hello @margud ,

I have checked this PR and LGTM.
It's QA approved :heavy_check_mark: 

Thank you !

**Comment by jolelievre** [comment]
Thanks @margud and @MhiriFaten

**Pull request merged**
